### ISSUE TITLE
Smart indentation bug

### ISSUE THREAD

**Issue by jkpubsrc** [mono/monodevelop#4307]
See here:

```csharp
private IEnumerable<Token[]> Group(IEnumerable<Token> tokens)
{
	List<Token> line = new List<Token>();
	bool bLastLineWasEmpty = true;
	foreach (Token t in tokens) {
		if (t.type.Equals("eol")) {
			if (line.Count > 0) {
				if (line[line.Count - 1].type.Equals("spc"))
					line.RemoveAt(line.Count - 1);
			}
			if (line.Count == 0) {
				if (!bLastLineWasEmpty) {
					yield return line.ToArray();
					bLastLineWasEmpty = true;
				}
			} else {
				yield return line.ToArray();
				bLastLineWasEmpty = false;
			}
			line.Clear();
		} else
		if (t.type.Equals("eof")) {
			if (line.Count > 0) {
				if (line[line.Count - 1].type.Equals("spc"))
					line.RemoveAt(line.Count - 1);
			}
				if (line.Count == 0) {
				if (!bLastLineWasEmpty) {
					yield return line.ToArray();
					bLastLineWasEmpty = true;
				}
			} else {
				yield return line.ToArray();
				bLastLineWasEmpty = false;
			}
			break;
		} else {
			line.Add(t);
		}
	}
}
```

Regardless that this might not be the most beautiful piece of code anyway, smart indentation has a bug here. See at about 2/3 of the snipplet at the second "`if (line.Count == 0) {`" : If I place the cursor before the `if` and press `backspace` the if-part is moved right to the preceding  `}`. If I then press enter again (where smart indentation is invoked) the indentation places the `ìf` as seen in the code snipplet above. Which is a bug as the `if` should be place one tab more left.

Please note that the first "`if (line.Count == 0) {`" works perfectly.

I'm using the following settings:

* An edited variant of "Visual Studio" named "Custom" with:
    * Indent block contents : on
    * Indent opened and closed braces : off
    * Indent switch sections : on
    * Indent case sections : on
    * Label indentation : one indent less

> VS bug [#591654](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/591654)

**Comment by xamarin-release-manager** [comment]
Fixed in version 7.8.0.286 (master)

Pull Request #5293 merged by: Lluis Sanchez
Author: mono
Commit: 4edf53d2399bf2e6e79b1d49160a066cd7f1de55 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
Fixed in version 7.7.0.1140 (release-7.7)

Pull Request #5293 merged by: Lluis Sanchez
Author: mono
Commit: 4edf53d2399bf2e6e79b1d49160a066cd7f1de55 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
Fixed in version 8.6.0.802 (master)

Pull Request #5293 merged by: Lluis Sanchez
Author: mono
Commit: 4edf53d2399bf2e6e79b1d49160a066cd7f1de55 (xamarin/monodevelop)
